File: mac/Classes/Plugins/PluginManager.m

#import "PluginManager.h"
#import "Plugin.h"
#import "Compiler.h"
#import "ActionType.h"
#import "ATFunctionalStyle.h"


static PluginManager *sharedPluginManager;


@implementation PluginManager {
    NSMutableDictionary *_actionTypesByIdentifier;
    NSMutableArray *_actionTypes;
}

@synthesize userPluginNames=_userPluginNames;

+ (PluginManager *)sharedPluginManager {
    if (sharedPluginManager == nil) {
        sharedPluginManager = [[PluginManager alloc] init];
    }
    return sharedPluginManager;
}

- (id)init {
    self = [super init];
    if (self) {
        _loadedPluginNames = [[NSMutableSet alloc] init];
    }

    return self;
}


- (void)loadPluginFromFolder:(NSString *)pluginFolder into:(NSMutableArray *)plugins {
    NSString *name = [[pluginFolder lastPathComponent] stringByDeletingPathExtension];

    if ([_loadedPluginNames containsObject:name])
        return;
    [_loadedPluginNames addObject:name];

    [plugins addObject:[[Plugin alloc] initWithPath:pluginFolder]];
}

- (void)loadPluginsFromFolder:(NSString *)pluginsFolder into:(NSMutableArray *)plugins {
    for (NSString *fileName in [[[NSFileManager defaultManager] contentsOfDirectoryAtPath:pluginsFolder error:nil] sortedArrayUsingSelector:@selector(compare:)]) {
        if ([[fileName pathExtension] isEqualToString:@"lrplugin"]) {
            [self loadPluginFromFolder:[pluginsFolder stringByAppendingPathComponent:fileName] into:plugins];
        }
    }
}

- (void)addActionType:(ActionType *)actionType {
    NSString *identifier = actionType.identifier;
    if (_actionTypesByIdentifier[identifier])
        return;

    if (actionType.valid) {
        [_actionTypes addObject:actionType];
        _actionTypesByIdentifier[identifier] = actionType;
    } else {
        NSLog(@"Skipped invalid action type def: %@", identifier);
    }
}

- (void)reloadPlugins {
    _actionTypes = [NSMutableArray new];
    _actionTypesByIdentifier = [NSMutableDictionary new];

    NSMutableArray *plugins = [NSMutableArray array];

    NSArray *libraryFolderPaths = [NSArray arrayWithObjects:@"~/Library/LiveReload", @"~/Dropbox/Library/LiveReload", nil];
    for (NSString *libraryFolderPath in libraryFolderPaths) {
        NSString *pluginsFolder = [[libraryFolderPath stringByAppendingPathComponent:@"Plugins"] stringByExpandingTildeInPath];
        [self loadPluginsFromFolder:pluginsFolder into:plugins];
    }

    _userPluginNames = [[NSArray alloc] initWithArray:[_loadedPluginNames allObjects]];

    NSString *bundledPluginsFolder;
    const char *pluginsOverrideFolder = getenv("LRBundledPluginsOverride");
    if (pluginsOverrideFolder && *pluginsOverrideFolder) {
        bundledPluginsFolder = [[NSString stringWithUTF8String:pluginsOverrideFolder] stringByExpandingTildeInPath];
    } else {
        bundledPluginsFolder = [[NSBundle mainBundle] resourcePath];
    }

    [self loadPluginsFromFolder:bundledPluginsFolder into:plugins];
    _plugins = [plugins copy];

    for (Plugin *plugin in _plugins) {
        for (ActionType *item in plugin.actionTypes) {
            [self addActionType:item];
        }
    }

    NSDictionary *pluginNamesByPath = [[(NSArray *)[_plugins valueForKeyPath:@"path"] dictionaryWithElementsGroupedIntoArraysByKeyPath:@"stringByDeletingLastPathComponent"] dictionaryByMappingValuesToKeyPath:@"lastPathComponent"];

    NSLog(@"Plugins loaded: %@", pluginNamesByPath);
    NSLog(@"Action types: %@", _actionTypes);

    NSArray *badPlugins = [_plugins filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"errors[SIZE] > 0"]];
    NSLog(@"Number of plugins with errors: %d", (int)badPlugins.count);
    for (Plugin *plugin in badPlugins) {
        NSLog(@"Error messages for %@:\n%@", plugin.path.lastPathComponent, plugin.errors);
    }
}

- (NSArray *)plugins {
    NSAssert(_plugins != nil, @"Plugins not loaded yet");
    return _plugins;
}

- (NSArray *)compilers {
    return [self valueForKeyPath:@"plugins.@unionOfArrays.compilers"];
}

- (NSArray *)compilerSourceExtensions {
    return [self valueForKeyPath:@"compilers.@unionOfArrays.extensions"];
}

- (Compiler *)compilerForExtension:(NSString *)extension {
    for (Compiler *compiler in self.compilers) {
        if ([compiler.extensions containsObject:extension]) {
            return compiler;
        }
    }
    return nil;
}

- (Compiler *)compilerWithUniqueId:(NSString *)uniqueId {
    for (Compiler *compiler in self.compilers) {
        if ([compiler.uniqueId isEqualToString:uniqueId]) {
            return compiler;
        }
    }
    return nil;
}

- (ActionType *)actionTypeWithIdentifier:(NSString *)identifier {
    return _actionTypesByIdentifier[identifier];
}

@end

File: mac/Classes/Plugins/PluginManager.swift

import Foundation
import AppKit

let g_sharedPluginManager = PluginManager()

class PluginManager : NSObject {

    class func sharedPluginManager() -> PluginManager {
        return g_sharedPluginManager
    }

    func reloadPlugins() {
        _actionTypes.removeAll()
        _plugins = []

        let libraryFolderPaths = ["~/Library/LiveReload", "~/Dropbox/Library/LiveReload"]
        for libraryFolderPath in libraryFolderPaths {
            let pluginsFolder = libraryFolderPath.stringByAppendingPathComponent("Plugins").stringByExpandingTildeInPath
            _loadPluginsFromFolder(pluginsFolder)
        }

        var bundledPluginsFolder = NSBundle.mainBundle().resourcePath!
        if let pluginsOverrideFolder = NSProcessInfo.processInfo().environment["LRBundledPluginsOverride"] as? String {
            let trimmed = pluginsOverrideFolder.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceAndNewlineCharacterSet())
            if !trimmed.isEmpty {
                bundledPluginsFolder = trimmed.stringByExpandingTildeInPath
            }
        }
        _loadPluginsFromFolder(bundledPluginsFolder)

        for plugin in _plugins {
            for actionType in plugin.actionTypes as ActionType[] {
                _addActionType(actionType)
            }
        }

        let badPlugins = _plugins.filter { !$0.errors.isEmpty }
        if !badPlugins.isEmpty {
            var errorMessage = ""
            errorMessage += "Number of plugins with errors: \(badPlugins.count)\n\n"
            for plugin in badPlugins {
                errorMessage += "Error messages for \(plugin.path.lastPathComponent):\n"
                for error in plugin.errors as NSError[] {
                    errorMessage += "• \(error.localizedDescription)\n"
                }
            }

            NSLog("%@", errorMessage)
            let alert = NSAlert()
            alert.messageText = "LiveReload couldn't load some plugins"
            alert.informativeText = errorMessage
            alert.addButtonWithTitle("Continue")
            alert.addButtonWithTitle("Quit")
            if alert.runModal() == NSAlertSecondButtonReturn {
                exit(1);
            }
        }

        _pluginsLoaded = true
    }

    var plugins: Plugin[] {
        assert(_pluginsLoaded, "Plugins not loaded yet")
        return _plugins
    }

    var _plugins: Plugin[] = []

    var compilers: Compiler[] {
        return flatten(_plugins.map { $0.compilers as Compiler[] })
    }

    var compilerSourceExtensions: String[] {
        let compiler = Compiler()
        return flatten(compilers.map { $0.extensions as String[] })
    }

    var userPluginNames: String[] {
        return String[](_loadedPluginNames.keys)
    }

    var actionTypes: ActionType[] {
        return _actionTypes.list
    }

    func compilerForExtension(ext: String) -> Compiler? {
        return findIf(compilers) { $0.usesExtension(ext) }
    }

    func compilerWithUniqueId(uniqueId: String) -> Compiler? {
        return findIf(compilers) { $0.uniqueId == uniqueId }
    }

    func actionTypeWithIdentifier(identifier: String) -> ActionType? {
        return _actionTypes[identifier]
    }

// private

    var _loadedPluginNames : Dictionary<String, Plugin> = [:]
    var _pluginsLoaded = false
    var _actionTypes = IndexedArray<String, ActionType>() { $0.identifier }

    func _loadPluginFromFolder(pluginFolder: String) {
        let name = pluginFolder.lastPathComponent.stringByDeletingPathExtension
        if !_loadedPluginNames[name] {
            let plugin = Plugin(path: pluginFolder)
            _loadedPluginNames[name] = plugin
            _plugins.append(plugin)
        }
    }

    func _loadPluginsFromFolder(pluginsFolder: String) {
        if let fileNames = NSFileManager.defaultManager().contentsOfDirectoryAtPath(pluginsFolder, error: nil) as String[]? {
            for fileName in fileNames {
                if fileName.pathExtension == "lrplugin" {
                    _loadPluginFromFolder(pluginsFolder.stringByAppendingPathComponent(fileName))
                }
            }
        }
    }

    func _addActionType(actionType: ActionType) {
        let identifier = actionType.identifier
        if actionType.valid {
            _actionTypes.append(actionType)
        } else {
            NSLog("Skipped invalid rule type def: \(identifier)")
        }
    }

}
